homeassistant/components/calendar/todoist.py

```diff
@@ -0,0 +1,604 @@
+                               data=task_json,
+                               headers={
+                                   "Content-Type": "application/json",
+                                   "X-Request-Id": str(uuid.uuid4()),
+                               })
+        if output.status_code == 200:
+            _LOGGER.info("Successfully posted task to Todoist.")
+        else:
+            error = "Could not post task to Todoist! "
+            error += "Status code: " + str(output.status_code)
+            _LOGGER.critical(error)
+    hass.services.register(DOMAIN, SERVICE_NEW_TASK, handle_new_task,
+                           descriptions[DOMAIN][SERVICE_NEW_TASK],
+                           schema=NEW_TASK_SERVICE_SCHEMA)
+
+
+class TodoistProjectDevice(CalendarEventDevice):
+    """A device for getting the next Task from a Todoist Project."""
+
+    def __init__(self, hass, data, labels, token,
+                 latest_task_due_date=None, whitelisted_labels=None,
+                 whitelisted_projects=None):
+        """Create the Todoist Calendar Event Device."""
+        self.data = TodoistProjectData(
+            data, labels, token, latest_task_due_date,
+            whitelisted_labels, whitelisted_projects
+        )
+
+        # Set up the calendar side of things
+        calendar_format = {}
+        calendar_format[CONF_NAME] = data['name']
+        calendar_format[CONF_DEVICE_ID] = data['name']
+
+        super().__init__(hass, calendar_format)
+
+    def update(self):
+        """Update all Todoist Calendars."""
+        # Set basic calendar data
+        super().update()
+
+        # Set Todoist-specific data that can't easily be grabbed
+        if self.data.event is not None and 'all_day' in self.data.event:
+            # This event doesn't have an end time, so we mark it as all day.
+            self._cal_data['all_day'] = True
+
+        self._cal_data['all_tasks'] = []
+        for task in self.data.all_project_tasks:
+            self._cal_data['all_tasks'].append(task['summary'])
+
+    def cleanup(self):
+        """Clean up all calendar data."""
+        super().cleanup()
+        self._cal_data['all_tasks'] = []
+
+    @property
+    def device_state_attributes(self):
+        """Return the device state attributes."""
+        if self.data.event is None:
+            # No tasks, we don't REALLY need to show anything.
+            return {}
+
+        attributes = super().device_state_attributes
+
+        # Add additional attributes.
+        attributes['due_today'] = self.data.event['due_today']
+        attributes['overdue'] = self.data.event['overdue']
+        attributes['all_tasks'] = self._cal_data['all_tasks']
+        attributes['priority'] = self.data.event['priority']
+        attributes['task_comments'] = self.data.event['comments']
+        attributes['task_labels'] = self.data.event['labels']
+
+        return attributes
+
+
+class TodoistProjectData(object):
```

Comment:
This should probably be broken out into a python package and published on pypi. Is there no pypi package already for todoist?

Reply:
[There is](https://pypi.python.org/pypi/todoist-python), but I didn't want to add any additional dependencies to the project, since Todoist has a REST API which accomplishes the same exact tasks. I considered adding their API as a dependency when creating the "Add Task" service, but in the end I felt that it wasn't worth it. When I was designing this, I had a misconception that using the API *required* you [to use their OAuth implementation](https://developer.todoist.com/sync/v7/#oauth), which would make my life much more difficult in terms of authenticating the user. That was another thing which pushed me towards using the REST API, although now that I'm looking back at it I see there's actually a way to authenticate with tokens after all.

That being said, half the code highlighted here is just me overriding HA's default Calendar implementation. If I were to move to the PyPI package, most of this code would largely be the same (since it's modifying the Todoist "device" based on the `TodoistProjectData` and not actually doing any real logic of its own). I would probably be able to get rid of the `TodoistProjectData` object by moving to PyPI, but I don't want to introduce more dependencies than I have to, especially when there's an "easy" way around it. I would also still need to "score" each task in order to display them properly in Home Assistant, something which is currently done in `TodoistProjectData`. If you'd like, though, I'm okay with doing that.

Reply:
[There is](https://pypi.python.org/pypi/todoist-python), but I didn't want to introduce any dependencies. When I first started designing the platform, I had the misconception after looking at [their Python API](https://developer.todoist.com/sync/v7/) that I would need to do everything via OAuth, which is obviously more of a challenge than just having the user specify a token in their config. Looking back at it now, I see that there is actually a way to do things with a token, but I'm still adverse to introducing extra dependencies.

That being said, I would still need *some* Todoist logic somewhere. This class has the scoring algorithm, which could be moved to the actual device. But (more importantly), it also supports "custom" projects -- i.e., where the user can ask to grab everything in the next week, or all things tagged "Homework", or what have you. That stuff comes from the config file, and it'll all have to be kept track of locally (otherwise users may wind up accidentally creating a new project as soon as the API syncs, from what I've gathered in a cursory examination).

This class also "converts" the data from Todoist to something that Home Assistant can process easier -- as an example, the due date logic around line 400 or so. Sure, I'd be able to get an actual datetime object from the API instead of relying upon `strptime`, but I would still need to "translate" that to the end time for a Calendar in Home Assistant. This also ensures that every single task has a start and end date, that there's always a list of comments (even if it's empty), etc. By default, if something is empty, there's just no entry for it in the API. The idea is to do as much logic as makes sense here, in order to cut down on the amount of logic contained within the actual device. I can probably move the logic for determining if something is due today or overdue into the device, but at the time it seemed fitting just to do it here, since I was already working with the data.

The end result is I would need some class which looks vaguely like this, except I would have references to Todoist's PyPI API calls instead of using `requests`. I'd still need to "score" each task, since Todoist sorts by date added, not by actual priority. I'd still need to "translate" a lot of the Todoist-specific things to what the Calendar device expects, but perhaps this can be done on the device end. Most importantly, though, I'd still need to keep a list of "local" projects for those user-defined devices that can be specified in the config. That'd be where most of the logic would come from.

I can do it, if you'd like me to. Doing so would give me access to Todoist Reminders, so I could probably expose the Reminder functionality built-in to Todoist to Home Assistant (right now, I can't access Reminders via the REST API -- just Tasks and their due dates). It would be a fairly easy change overall, but I'm not sure whether it would cut down on the amount of platform-specific logic unless I got rid of the custom user-defined projects (which is where a lot of the logic comes from currently).

Reply:
Introducing a new dependency is not a problem, but rather part of the ground rules for home assistant:
https://home-assistant.io/developers/code_review_platform/#6-communication-with-devicesservices
https://home-assistant.io/developers/code_review_platform/#1-requirements

Having an interface class or callback function between the pypi package API and the home assistant entity API is ok, but we try to minimize it as much as possible. Usually this class/function is responsible for fetching/accepting new data into home assistant.